### PR TITLE
Belize theme

### PR DESCRIPTION
**Pull request by flaiker**
#2610 
#2700 

@sbcgua

I added a minimal CSS variable polyfill implementation and set the default theme to always load and add others on top. That should reduce the duplication quite a lot and add the possibility to refactor some things.

I can remove the variable stuff again if you don't like it.

I have not tested the abapmerge version yet.

### CODE REVIEW

**Comment by flaiker** [comment]
I am adding some new colors to the theme. Seems like this is the reference page:
https://experience.sap.com/fiori-design-web/colors/
Though not many exact colors from there were used in the GUI Belize theme it seems like.

**Comment by flaiker** [comment]
Hmm, theme switching doesn't really work right now as the CSS variables are evaluated on asset registration and not when the CSS is actually loaded. This leads to the belize variables always being loaded regardless of if the CSS was actually included.
One could not register the themes that are not in use. Then you have to restart abapGit on theme change though (which shouldn't be that bad?).
The cleaner way would be to evaluate the variables client side but I don't have confidence in implementing that.

**Comment by sbcgua** [comment]
Hmmm ... a lot to discuss :)

1) Variables are cool ... i personally still think it is an overcomplecation, but it is cool :) :+1:
2) But I think the place is not ideal (assetman). Preprocessor should be certainly a separete class and, I'd say, should be invoked from a page, not on global asset load.
3) Also, as you mentioned, assets are cached on start. This is something I wanted to change but didn't have time :) The assetman->register_asset has a flag - IV_CACHABLE - which makes GUI to skip the caching, but then there is no normal mechanism to activate an asset later from a page. I had an idea to pass gui and asset man to a component's constructor or `render` so that the component can load specific assets and cache them if needed, but didn't finish that code.

In my opinion the code should look something like this (pseudo code):

```
@ UI_FACTORY
...
register_asset( name = 'theme_x', iv_cachable = false )
...

@ main_page->render( gui_instance, assetman_instance )
...
main_css_src = assetman_instance->get_asset( 'common.css' )
theme_css_src = assetman_instance->get_asset( 'theme_x.css' )

preprocessor = new cool_abap_css_preprocessor( )
css_content = preprocessor->preprocess( src = main_css_src, mixin = theme_css_src ) 

  // where mixin is an optional param
  // the PP merges 2 files, 
  // and then detects/applies variables REPLACING earlier ones with later ones (!)
  // which gives the desired effect for the themes

gui_instance->cache_asset( name = 'common.css', content = css_content )

```

What do you think on this ?

P.S. "...evaluate the variables client side..." - what do you mean? JS ? If yes, imho preprocessing is more universal - JS does not work well in java clients

P.P.S. imho the quick way to deliver this feature is to postpone variable for now ... on the other hand we are not in rush, are we ? :) And variables are cool :)

**Comment by flaiker** [comment]
> In my opinion the code should look something like this (pseudo code):
> 
> ```
> @ UI_FACTORY
> ...
> register_asset( name = 'theme_x', iv_cachable = false )
> ...
> 
> @ main_page->render( gui_instance, assetman_instance )
> ...
> main_css_src = assetman_instance->get_asset( 'common.css' )
> theme_css_src = assetman_instance->get_asset( 'theme_x.css' )
> 
> preprocessor = new cool_abap_css_preprocessor( )
> css_content = preprocessor->preprocess( src = main_css_src, mixin = theme_css_src ) 
> 
>   // where mixin is an optional param
>   // the PP merges 2 files, 
>   // and then detects/applies variables REPLACING earlier ones with later ones (!)
>   // which gives the desired effect for the themes
> 
> gui_instance->cache_asset( name = 'common.css', content = css_content )
> ```
> 
> What do you think on this ?

That makes sense to me and is a better solution than what's currently in the PR. 👍 

> P.S. "...evaluate the variables client side..." - what do you mean? JS ? If yes, imho preprocessing is more universal - JS does not work well in java clients

Well, we could just add the [minified](https://cdn.jsdelivr.net/npm/css-vars-ponyfill@2) distribution file of [jhildenbiddle/css-vars-ponyfill](https://github.com/jhildenbiddle/css-vars-ponyfill) as W3MI and call it a day. That gets us a maintained and tested polyfill for css variables instead of a selfmade preprocessor. One benefit would also be that the page will load normally if you do a "Save as..." html, get the CSS and JS files and open it normally in a browser for debugging purposes. The cost would be an external dependency (and maybe worse performance?).

> P.P.S. imho the quick way to deliver this feature is to postpone variable for now ... on the other hand we are not in rush, are we ? :) And variables are cool :)

Sure, I just would like to not leave a maintenance mountain behind me for the next one to struggle with when adding some new CSS classes. And without variables you would have to write some comments in the theme file advising people to do search and replace on these values I guess? So if we can find an easy way for variables we should try imo.

Let me know what you think about that JS library. Haven't actually tested it yet with SAP GUI but it looks quite promising to me.

**Comment by larshp** [comment]
would like to avoid adding external dependencies if possible

**Comment by flaiker** [comment]
I moved stuff around a bit according to @sbcgua 's suggestions (I hope I understood them correctly).

File: src/ui/zcl_abapgit_gui_css_processor.clas.abap [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    " 3. Replace all other variable usages by inlining the values.
+    LOOP AT lt_contents ASSIGNING <lv_content>.
+      LOOP AT lt_css_variables ASSIGNING <ls_css_variable>.
+        REPLACE ALL OCCURRENCES OF |var(--{ <ls_css_variable>-name })|
+                IN <lv_content>
+                WITH <ls_css_variable>-value.
+      ENDLOOP.
+    ENDLOOP.
+
+    rv_result = concat_lines_of( table = lt_contents sep = cl_abap_char_utilities=>cr_lf ).
```
Comment (sbcgua) [inline src/ui/zcl_abapgit_gui_css_processor.clas.abap]: maybe newline better ? mostly newline char is used over the AG files from what I remember

**Comment by sbcgua** [comment]
... Thinking more about future preprocessor improvements: I think it would be really nice to make it automatic. So that it detects css resources directly from html and merges them. Then: a) debug is straight forward b) less direct code c) preprocessor can become ui_core for easy reuse in other projects (BTW current code breaks the reusability of UI_CORE) ...

File: src/ui/zcl_abapgit_gui_css_processor.clas.abap (hunk 1 above)
[the remark this answers is not in the document]
Comment (flaiker) [reply src/ui/zcl_abapgit_gui_css_processor.clas.abap]: Yep, changed it

**Comment by flaiker** [comment]
I added some tests. I don't think it is necessary to have it run over the whole html code as styling and variable usage _should_ be done only in the CSS files anyway and referenced from HTML using classes. For it to be automatic it would of course be quite nice.

I didn't think of the separation of UI_CORE. I guess it should not know of the abapGit-specific CSS files right? We can move css_processor to UI_CORE and add a method to GUI to specify the CSS files from the outside? I am not that familiar with the architecture of the UI. Maybe that can be a follow-up PR.

I'll fix the linter issues.

**Comment by sbcgua** [comment]
:+1: 
I cannot test the code now, but from screening perspective it looks ok and can be merged. 

>> process whole html

No, i meant just extracting css references and replace them with a processed ones. So that explicit ‘add’ is not necessary. But for sure a separate PR. 

>> specific names in core_ui 

Yes :) that was my point. The above auto processing should solve it. also, for sure, in another PR.

**Comment by larshp** [comment]
@sbcgua will you be performing a small test?

**Comment by sbcgua** [comment]
Yeap, sorry for some silence, had been busy with other stuff.
Installed and tested - works fine for me.
@flaiker by the way really fresher look, great work, thanks !

**Pull request merged**